doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+
+```
+ptvc' = [ptvc-updates ⋃ { u = [data, time, diff] in tvc_updates |
+                          (upper <=t time) and not ( upper' <=t time)}
+         since,
+         upper']
+```
+
+Is `ptvc'` correct? Because `since` is unchanged and its updates between
+`since` and `upper` are unchanged, it must still be correct for any `t` between
+`since` and `upper`. And since it now includes every TVC update between `upper`
+and `upper'`, it must also be correct for those times as well: their
+`prior-updates` are identical, by induction to times which were correct in the
+original `ptvc`. `ptvc'` is therefore correct for all times between `since` and
+`upper'`.
+
+This allows us to advance the upper frontier of a pTVC by appending new updates
+to it (so long as we have every update between `upper` and `upper'`) and still
+preserve correctness.
+
+### pTVC compaction
+
+Again, let `ptvc = [ptvc-updates, since, upper]` be a correct view of some TVC.
+Now consider compacting all updates prior to some frontier `since'`, in such a
+way that the collection versions read at times later than or equal to `since'`
+remain unchanged. We write this `ptvc' = compact(ptvc, since')`:
+
+```
+updates' = compact(updates, since')
+ptvc' = [updates', since', upper]
+```
+
+Since `updates'` has been compacted to `since'`, it contains no updates with
+times prior to or concurrent with `since'`. Presumably, `updates'` differs from
+`updates` only along the "boundary layer"--the times along the leading edge of
+the set of times which are at or later than `since'`. Updates prior to or
+concurrent with `since'` have been merged forward into that boundary layer,
+preserving correctness.
+
+How exactly compaction works is described in [K-Pg: Shared State in
+Differential
+Dataflows](https://github.com/TimelyDataflow/differential-dataflow/blob/master/sigmod2019-submission.pdf)
+
+### Time evolution of pTVCs
+
+Materialize's goal is to keep track of a window of each relevant TVC over time.
+At any given moment a TVC is represented by a single `ptvc = [updates, since,
+upper]` triple. Each TVC begins with the empty `ptvc0`, which is a correct
+representation of every TVC. As new information about the TVC becomes
+available, Materialize appends those updates using `append(ptvc, upper',
+new-updates)`. To reclaim storage space, Materialize compacts ptvcs using
+`compact(ptvc, since')`. Since both `append` and `compact` preserve pTVC
+correctness, Materializ always allows us to query a range of times (those
+between the current `since` and `upper`), and obtain the same results as would
+be observed in the "real" TVC.
+
+During this process the `since` and `upper` frontiers never move
+backwards---and typically, they advance.
+
+## The TVC Map
+
+To track these changing pTVCs over time, Materialize maintains a mutable *TVC
+map*. The keys in this map are `GlobalId`s, and the values are pTVCs. Each TVC
+has at most one `GlobalID` over the entire history of a Materialize cluster. In
```

Comment:
I think "Each TVC has at most one `GlobalId` ..." might want to be "Each `GlobalId` has at most one TVC ..."? What do you think?

Reply:
I honestly don't know--this one's up to you! Is the goal to have multiple concurrent pTVCs, each with their own GlobalId, which correspond to a single TVC? I think this probably connects with whether you want the adapter's "source" to refer to TVCs or pTVCs too.

Reply:
I guess I'm reading "Each TVC has at most one `GlobalId`" to imply by omission that "a `GlobalId` can have multiple TVCs", which wouldn't make sense to me.

We can chew on this. My mental model is:
1. Materialize models changing data through an assignment to each `GlobalId` of *at most one* TVC.
2. The realization of that model assigns, at any moment, to each `GlobalId` a pTVC that is always compatible with the modeled TVC.

My hunch is that the same words mean different things to us, or that we are focusing on different implications of them, rather than that we disagree. I think.

Reply:
I'm not disagreeing, haha--I literally don't know how Materialize is supposed to work, and I can see lots of defensible options. Let me try rewriting this based on your comment, and see what you think.